File: android/app/src/main/java/com/tappyai/app/chat/ShoppingDecisionView.kt
package com.tappyai.app.chat

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull

/**
 * D1 — the shopping DECISION, as it arrives inside `[TAPPY_SHOPPING]`.
 *
 * Kotlin mirror of web's `SynthesisView` (src/lib/ai/consultative/synthesisView.ts). The server
 * serialises that view directly into the marker, so this is the wire shape, not a re-derivation:
 * the config label, price range, match verdict and recommended flag are EXACTLY what the model was
 * told, which is why the card can never disagree with the prose above it.
 *
 * Android decoded none of this before. The block was stripped and thrown away, so a mobile user on
 * a shopping turn read the prose and silently lost the entire decision — the recommendation, the
 * price ranges, the match verdicts and every alternative. That is the largest functional parity gap
 * the V3 audit found (V3_PLATFORM_PARITY.md D1).
 *
 * 🚨 THE HONESTY RULE IS IN THE TYPES. Every field the server may not know is nullable, and the UI
 * renders null as an explicit "chưa rõ". Nothing here is defaulted to a plausible-looking value,
 * because a fabricated price is worse than an absent one. `buildSynthesisView` on the server groups
 * and ranks; this model and its card group NOTHING and infer NOTHING.
 */
@Serializable
data class ShoppingOfferView(
    val seller: String? = null,
    val url: String? = null,
    val price: Double? = null,
    val currency: String? = null,
    val condition: String? = null,
    /**
     * The rating THIS listing carries, never the entity's — Serper puts it on the row, so it
     * belongs to one seller's listing (web `SynthesisOfferView.rating`). Was not decoded before:
     * the wire said `rating 4.7 · 339` and the phone showed neither (UAT 2026-09-12).
     */
    val rating: Double? = null,
    @SerialName("ratingCount") val ratingCount: Int? = null,
)

/** One stated specification: the listing's own figure, never derived (web `SynthesisSpecView`). */
@Serializable
data class ShoppingSpecView(
    /** `chip` | `ram` | `storage` | `size`. */
    val key: String = "",
    /** The listing's figure — a number or a string on the wire, read as text. */
    val value: JsonPrimitive? = null,
) {
    val valueText: String? get() = value?.contentOrNull?.takeIf { it.isNotBlank() }
}

/** The seller's condition wording plus a key a dictionary can translate (web `condition`). */
@Serializable
data class ShoppingConditionView(
    val key: String? = null,
    val label: String = "",
)

/** One product configuration the assistant considered. */
@Serializable
data class ShoppingEntityView(
    val key: String = "",
    val config: String = "",
    /** `khop` = matches the request · `khac` = differs · `chua_ro` = not enough information. */
    @SerialName("matchesRequest") val matchesRequest: String = "chua_ro",
    val recommended: Boolean = false,
    @SerialName("priceLow") val priceLow: Double? = null,
    @SerialName("priceHigh") val priceHigh: Double? = null,
    /** A representative product photo, if any offer carried one. */
    val image: String? = null,
    val offers: List<ShoppingOfferView> = emptyList(),
    /** The product's own name (the listing title), when the server projected it. LAST so the
     *  positional call sites that predate it keep compiling; the card reads [displayName]. */
    val name: String? = null,
    /** The stated configuration as data, so the card can label each part (web `specs`). */
    val specs: List<ShoppingSpecView> = emptyList(),
    val condition: ShoppingConditionView? = null,
) {
    val displayName: String? get() = name?.trim()?.takeIf { it.isNotEmpty() }
}

/** Why the server recommends what it recommends. A recommendation without reasons is not shown. */
@Serializable
data class ShoppingRecommendationView(
    @SerialName("entityKey") val entityKey: String? = null,
    val seller: String? = null,
    val reasons: List<ShoppingReason> = emptyList(),
    @SerialName("tradeOff") val tradeOff: ShoppingReason? = null,
    val conditional: Boolean = false,
)

@Serializable
data class ShoppingReason(
    val attribute: String = "",
    /** The engine's own (English) wording — the fallback when the client has no sentence for it. */
    val evidence: String = "",
    /**
     * The data behind the sentence (`value`, `count`, `priceVnd`, `km`, `stars`, `minutes`), so the
     * client can SAY it in the user's language (web `reasonText`). Dropped before, which is why the
     * phone printed `rated 4.7 / 95000 VND / 339 reviews` under a Vietnamese reply.
     */
    val params: Map<String, JsonPrimitive>? = null,
) {
    fun number(name: String): Double? = params?.get(name)?.doubleOrNull
}

@Serializable
data class ShoppingDecisionView(
    val v: Int = 1,
    val entities: List<ShoppingEntityView> = emptyList(),
    val recommendation: ShoppingRecommendationView? = null,
    /**
     * The configuration the USER asked for, or `null` when they named none (web `requested`).
     *
     * A match verdict is meaningless without a request to match: `chua_ro` is emitted both when a
     * listing is unclear and when nothing was asked, so the badge is only offered when the request
     * is known. Kept as raw JSON so an OLD marker written before the field existed (key absent) is
     * told apart from an explicit `null` — web treats absent as "show" (`!== null`) and so does this.
     */
    val requested: JsonElement? = ABSENT,
) {
    /** Web: `showMatch = view.requested !== null`. */
    val showsMatchBadge: Boolean get() = requested === ABSENT || (requested != null && requested !is JsonNull)

    val requestedText: String? get() = (requested as? JsonPrimitive)?.contentOrNull
}

/**
 * Sentinel for "the key was not on the wire" (`ShoppingDecisionView.requested`). Top-level on
 * purpose: a PRIVATE companion on a `@Serializable` class hides the plugin-generated
 * `Companion.serializer()` from every other class, so `decodeFromString<ShoppingDecisionView>`
 * in the parser threw `IllegalAccessError` — swallowed by `runCatching` as "no decision" — while
 * the same call compiled inside the test source set worked (2026-09-12, 27 red tests).
 */
private val ABSENT: JsonElement = JsonPrimitive("__absent__")

/** Match verdicts the server can emit. Kept as a closed set so the UI cannot invent a fourth. */
object ShoppingMatch {
    const val EXACT = "khop"
    const val DIFFERENT = "khac"
    const val UNKNOWN = "chua_ro"
}

File: ios/TappyAI/Features/Chat/Model/ShoppingDecisionView.swift
import Foundation

/// D1 — the shopping DECISION, as it arrives inside `[TAPPY_SHOPPING]`.
///
/// Swift mirror of web's `SynthesisView` (src/lib/ai/consultative/synthesisView.ts) and of
/// Android's `ShoppingDecisionView.kt`. The server serialises that view straight into the marker,
/// so this is the wire shape rather than a re-derivation: the config label, price range, match
/// verdict and recommended flag are EXACTLY what the model was told, which is why the card can
/// never disagree with the prose above it.
///
/// iOS decoded none of this before — `stripMarkerResidue` removed the block and threw it away, so
/// a user on a shopping turn read the prose and silently lost the recommendation, the price
/// ranges, the match verdicts and every alternative (V3_PLATFORM_PARITY.md D1).
///
/// 🚨 THE HONESTY RULE IS IN THE TYPES. Every field the server may not know is optional, and the
/// view renders nil as an explicit "chưa rõ". Nothing is defaulted to a plausible-looking value,
/// because a fabricated price is worse than an absent one.

struct ShoppingOfferView: Codable, Equatable, Sendable {
    let seller: String?
    let url: String?
    let price: Double?
    let currency: String?
    let condition: String?
}

/// A Commerce Link the platform attached to a product entity (web `SynthesisCommerceView`,
/// src/lib/ai/consultative/synthesisView.ts; Android `ShoppingCommerceView`). The Shopping card
/// renders from the `[TAPPY_SHOPPING]` marker rather than the live place view, so the same
/// canonical handoff is projected here with the same facts: merchant, depth, login boundary, the
/// RESOLVED `labelKey` and the opaque ids the handoff beacon reports.
///
/// 🚨 iOS DECODED NONE OF THIS BEFORE. `commerceLinks` was an unknown key and was ignored, so
/// "Mua iPhone trên TikTok Shop" showed the web its TikTok Shop handoff and showed the phone a
/// Google Shopping redirect list. Same marker, two merchants — the drift the shared fixtures pin.
struct ShoppingCommerceView: Codable, Equatable, Sendable, Identifiable {
    var linkId: String = ""
    var requestId: String = ""
    var providerId: String = ""
    var merchantName: String = ""
    var url: String = ""
    var depth: Int = 0
    var guestDepth: Int = 0
    var authRequiredAt: String = ""
    var loginRequired: Bool = false
    var freshnessType: String = ""
    var expiresAt: String?
    var tracked: Bool = false
    var capability: String?
    var primary: Bool = true
    /// `SEARCH_HANDOFF` (the merchant's search for the user's words) vs a detail / checkout handoff.
    var kind: String = ""
    /// The RESOLVED label key. Nil on a marker written before the field existed (see `labelKeyOrFallback`).
    var labelKey: String?
    var handoff: String?
    var authenticatedDepth: Int?
    var facts: LiveCommerceObservedFacts?

    var id: String { linkId.isEmpty ? url : linkId }
    var isSearch: Bool { kind == "SEARCH_HANDOFF" }

    /// A marker persisted before the server projected `labelKey` still renders an honest label: a
    /// search says "search on", a detail handoff says "view on" — never a stronger verb than the
    /// server chose (the resolver on the server is the only thing allowed to promise a purchase).
    var labelKeyOrFallback: String { labelKey ?? (isSearch ? "v3.action.searchOn" : "v3.action.viewOn") }

    /// The same action shape the live place card renders, so one label resolver serves both.
    var asPlaceCardAction: PersistedPlaceAction {
        PersistedPlaceAction(
            kind: "purchase",
            urlKind: isSearch ? "search" : "direct",
            url: url,
            labelKey: labelKeyOrFallback,
            platform: merchantName,
            attributed: nil,
            commerce: LiveCommerceFacts(
                linkId: linkId, requestId: requestId, providerId: providerId, depth: depth, guestDepth: guestDepth,
                authRequiredAt: authRequiredAt, loginRequired: loginRequired, handoff: handoff,
                authenticatedDepth: authenticatedDepth, freshnessType: freshnessType, expiresAt: expiresAt,
                tracked: tracked, capability: capability, primary: primary, facts: facts
            )
        )
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        linkId = (try? c.decode(String.self, forKey: .linkId)) ?? ""
        requestId = (try? c.decode(String.self, forKey: .requestId)) ?? ""
        providerId = (try? c.decode(String.self, forKey: .providerId)) ?? ""
        merchantName = (try? c.decode(String.self, forKey: .merchantName)) ?? ""
        url = (try? c.decode(String.self, forKey: .url)) ?? ""
        depth = (try? c.decode(Int.self, forKey: .depth)) ?? 0
        guestDepth = (try? c.decode(Int.self, forKey: .guestDepth)) ?? 0
        authRequiredAt = (try? c.decode(String.self, forKey: .authRequiredAt)) ?? ""
        loginRequired = (try? c.decode(Bool.self, forKey: .loginRequired)) ?? false
        freshnessType = (try? c.decode(String.self, forKey: .freshnessType)) ?? ""
        expiresAt = try? c.decode(String.self, forKey: .expiresAt)
        tracked = (try? c.decode(Bool.self, forKey: .tracked)) ?? false
        capability = try? c.decode(String.self, forKey: .capability)
        primary = (try? c.decode(Bool.self, forKey: .primary)) ?? true
        kind = (try? c.decode(String.self, forKey: .kind)) ?? ""
        labelKey = try? c.decode(String.self, forKey: .labelKey)
        handoff = try? c.decode(String.self, forKey: .handoff)
        authenticatedDepth = try? c.decode(Int.self, forKey: .authenticatedDepth)
        facts = try? c.decode(LiveCommerceObservedFacts.self, forKey: .facts)
    }

    init(url: String, merchantName: String, kind: String, labelKey: String? = nil, linkId: String = "", requestId: String = "", providerId: String = "") {
        self.url = url
        self.merchantName = merchantName
        self.kind = kind
        self.labelKey = labelKey
        self.linkId = linkId
        self.requestId = requestId
        self.providerId = providerId
    }
}

/// The handoffs the card shows, split the way web splits them (`handoffsOf`,
/// components/chat/ShoppingDecision.tsx): DETAIL links are the buttons; the marketplaces' SEARCH
/// fallbacks are offered only when the entity has no detail link at all.
struct ShoppingCommerceHandoffs: Equatable, Sendable {
    let detail: [ShoppingCommerceView]
    let search: [ShoppingCommerceView]
    var isEmpty: Bool { detail.isEmpty && search.isEmpty }
}

/// One product configuration the assistant considered.
struct ShoppingEntityView: Codable, Equatable, Sendable, Identifiable {
    let key: String
    let config: String
    /// `khop` = matches the request · `khac` = differs · `chua_ro` = not enough information.
    let matchesRequest: String
    let recommended: Bool
    let priceLow: Double?
    let priceHigh: Double?
    /// A representative product photo, if any offer carried one.
    let image: String?
    let offers: [ShoppingOfferView]
    /// The entity's leading DETAIL handoff (a stored marker from before `commerceLinks` carries only this).
    let commerce: ShoppingCommerceView?
    /// Every Commerce Link CCP attached, in CCP's ranking order.
    let commerceLinks: [ShoppingCommerceView]?

    var id: String { key }

    var commerceHandoffs: ShoppingCommerceHandoffs {
        let all = commerceLinks ?? (commerce.map { [$0] } ?? [])
        let detail = Array(all.filter { !$0.isSearch }.prefix(5))
        let search = detail.isEmpty ? Array(all.filter { $0.isSearch }.prefix(2)) : []
        return ShoppingCommerceHandoffs(detail: detail, search: search)
    }

    /// Restores the memberwise initialiser that `init(from:)` below suppresses.
    init(
        key: String,
        config: String,
        matchesRequest: String,
        recommended: Bool = false,
        priceLow: Double? = nil,
        priceHigh: Double? = nil,
        image: String? = nil,
        offers: [ShoppingOfferView] = [],
        commerce: ShoppingCommerceView? = nil,
        commerceLinks: [ShoppingCommerceView]? = nil
    ) {
        self.key = key
        self.config = config
        self.matchesRequest = matchesRequest
        self.recommended = recommended
        self.priceLow = priceLow
        self.priceHigh = priceHigh
        self.image = image
        self.offers = offers
        self.commerce = commerce
        self.commerceLinks = commerceLinks
    }

    // Defaults on decode: a field the server omitted must not fail the whole decision, but it must
    // also never acquire a value. Optionals stay nil; only identity and the closed-set verdict get
    // a fallback, and the verdict falls back to "unknown" rather than to a claim.
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        key = (try? c.decode(String.self, forKey: .key)) ?? UUID().uuidString
        config = (try? c.decode(String.self, forKey: .config)) ?? ""
        matchesRequest = (try? c.decode(String.self, forKey: .matchesRequest)) ?? ShoppingMatch.unknown
        recommended = (try? c.decode(Bool.self, forKey: .recommended)) ?? false
        priceLow = try? c.decode(Double.self, forKey: .priceLow)
        priceHigh = try? c.decode(Double.self, forKey: .priceHigh)
        image = try? c.decode(String.self, forKey: .image)
        offers = (try? c.decode([ShoppingOfferView].self, forKey: .offers)) ?? []
        commerce = try? c.decode(ShoppingCommerceView.self, forKey: .commerce)
        commerceLinks = try? c.decode([ShoppingCommerceView].self, forKey: .commerceLinks)
    }
}

struct ShoppingReason: Codable, Equatable, Sendable {
    let attribute: String
    let evidence: String
}

/// Why the server recommends what it recommends. A recommendation without reasons is not shown.
struct ShoppingRecommendationView: Codable, Equatable, Sendable {
    let entityKey: String?
    let seller: String?
    let reasons: [ShoppingReason]
    let tradeOff: ShoppingReason?
    let conditional: Bool

    /// Restores the memberwise initialiser that `init(from:)` below suppresses.
    init(
        entityKey: String?,
        seller: String? = nil,
        reasons: [ShoppingReason] = [],
        tradeOff: ShoppingReason? = nil,
        conditional: Bool = false
    ) {
        self.entityKey = entityKey
        self.seller = seller
        self.reasons = reasons
        self.tradeOff = tradeOff
        self.conditional = conditional
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        entityKey = try? c.decode(String.self, forKey: .entityKey)
        seller = try? c.decode(String.self, forKey: .seller)
        reasons = (try? c.decode([ShoppingReason].self, forKey: .reasons)) ?? []
        tradeOff = try? c.decode(ShoppingReason.self, forKey: .tradeOff)
        conditional = (try? c.decode(Bool.self, forKey: .conditional)) ?? false
    }
}

struct ShoppingDecisionView: Codable, Equatable, Sendable {
    let v: Int
    let entities: [ShoppingEntityView]
    let recommendation: ShoppingRecommendationView?

    /// Declaring `init(from:)` below suppresses Swift's synthesised memberwise initialiser, so it
    /// is restored explicitly. Tests and previews construct these values directly.
    init(v: Int = 1, entities: [ShoppingEntityView], recommendation: ShoppingRecommendationView? = nil) {
        self.v = v
        self.entities = entities
        self.recommendation = recommendation
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        v = (try? c.decode(Int.self, forKey: .v)) ?? 1
        entities = (try? c.decode([ShoppingEntityView].self, forKey: .entities)) ?? []
        recommendation = try? c.decode(ShoppingRecommendationView.self, forKey: .recommendation)
    }
}

/// Match verdicts the server can emit. A closed set, so the UI cannot invent a fourth.
enum ShoppingMatch {
    static let exact = "khop"
    static let different = "khac"
    static let unknown = "chua_ro"
}

/// A price range, a single bound, or an explicit unknown.
///
/// Free function because this is where the honesty rule is easiest to break: an empty string, a
/// "0đ", or a low bound presented as a range would each be a claim the server never made.
func shoppingPriceRangeText(low: Double?, high: Double?, unknown: String) -> String {
    func compact(_ amount: Double) -> String {
        let v = Int(amount)
        if v >= 1_000_000 {
            let millions = v / 1_000_000
            let remainder = (v % 1_000_000) / 100_000
            return remainder > 0 ? "\(millions).\(remainder) triệu" : "\(millions) triệu"
        }
        return "\(v / 1000)k"
    }
    switch (low, high) {
    case let (l?, h?) where l != h: return "\(compact(l)) – \(compact(h))"
    case let (l?, _): return compact(l)
    case let (_, h?): return compact(h)
    default: return unknown
    }
}
